Design.md

```diff
@@ -181,7 +181,9 @@ This API creates a user on Vault.
 1. The `get-policy` API will be called with the `userPolicy@[account-id]` policy name using assumed role credentials
 1. If the `get-policy` API does not return any policy, using the assumed role credentials creates an IAM managed policy with the `userPolicy@[account-id]` policy name with full S3 access.
 1. `attach-policy` for the new user will be called using the assumed role credentials and the policy arn formatted as `arn:aws:iam::[account-id]:policy/userPolicy@[account-id]`.
-1. `create-s3credentials` api will be called using assumed role credentials
+1. The IAM `generate-access-key` API will be called using assumed role credentials.
+1. Encrypt the secret key using the AES algorithm. Store the encrypted secret key on Redis Cache in the hash named "OSIS", with the hash key formatted as `<Username>__<AccessKeyID>`. 
```

Comment:
Since data will be encrypted using a symmetric key (to be defined which block cipher mode to be used, by the way...), please make sure key rotation is taken into account from day 0. I believe I saw somewhere the key is to be stored in some file somewhere, fine, but please build the system such that multiple keys can be defined, they can be rotated over time, and which key to be used to decrypt can be retrieved from the data (meta-information next to the encrypted string).

Reply:
@NicolasT 
Sure. I will change the design to use `AES256GCM`. 

Regarding rotating keys management, please correct if my understanding is correct. 
* We will add multiple keys to the `application.properties` config file and same values will exist on all nodes running OSIS application.
* During encryption, we will use one of the keys _(key will be chosen in a round robin fashion)_ to encrypt the data.
* Concat the cipher text with metadata regarding the key used.
* During decryption, use the metadata to identify the key used and use the key from the `application.properties` to decrypt the cipher
* We will have limited number of keys _(Say 10, if not please suggest)_ and we shall never remove them from the config file on any OSIS node.
* We will encode the keys using Base64 before putting them on config files. So these keys will be decoded by OSIS before using them in AES encryption.

Reply:
> During encryption, we will use one of the keys (key will be chosen in a round robin fashion) to encrypt the data.

No, that's not what I meant by 'rotation'. Key rotation means you can replace a key with another one once the original key is either compromised, or there's some rule which dictates how long a key may be used. It's not about "Using a bunch of keys all at the same time".

> We will have limited number of keys (Say 10, if not please suggest) and we shall never remove them from the config file on any OSIS node.
The intent is exactly to be *able* to remove a key from the config file, after it's been rotate out.

Often this is implemented as follows: indeed, in some configuration, you keep some 'key slots', mapping a unique value (say, an incrementing natural) to a key and the encryption method used. So, at first:

```
keys:
  - id: 0
    cipher: AES256GCM
    key: abcdef
```
You take the 'first' key out of that list, use it to encrypt, and indeed keep "key 0 was used" in meta-information with the encrypted value. Now, say, the key was compromised. So, you change the configuration:
```
keys:
  - id: 1
    cipher: AESCBC+PKCS7
    secretKey: qwerty
  - id: 0
    cipher: AES256GCM
    secretKey: abcdef
```
From now on, all new data will be encrypted with the new key, using a stronger encryption scheme. Now, we need a procedure which will read all encrypted values from the datastore and decrypt them (at least of they use scheme `0`), then re-crypt them (now using scheme `1`), and store them again in the datastore, overwriting the original value.
Once all data has been recrypted, we can remove slot `0` from the config:
```
keys:
  - id: 1
    cipher: AESCBC+PKCS7
    secretKey: qwerty
```
If key `1` is not compromised, but we simply want to use another cipher for future data, without recrypting existing data, it can simply be added:
```
keys:
  - id: 2
    cipher: Secretbox
    secretKey: thisisareallylongandstrongkey
  - id: 1
    cipher: AESCBC+PKCS7
    secretKey: qwerty
```
-----
Also, consider Secretbox instead of AESGCM, which gives both encryption and MAC at once. And take care of properly using nonces, which should also go in the meta-information.

Reply:
@NicolasT 
So do we need to expose a new API or route to refresh the keys on Redis for re-crypting. Otherwise there is no other way we can ensure all the original data is re-crypted. 

Risk here is, if we miss any of the data to re-crypt or while any of the data failed re-crypting , then we will get errors on our application as it will not be able to decrypt as we do not have the original key which encrypted it.

Reply:
It's not necessarily some API or some route. Can simply be a utility/script which, given a configuration file, can so what's required against a Redis database. Key rotation is a 'triggered' procedure anyway (either manually, or automatically). The procedure is "Add key to config file(s), run this command to recrypt, if successful, remove old data from config file(s)". The recrypt operation is idempotent, so if it fails intermittently for some reason, one can re-run it.

Reply:
One more clarification: in the example above, I use multiple cipher methods. However, that does not mean we should *support* more than one method *today*. However, we should be able to move to another cipher in the future. Hence, we need to have the 'hooks' to do so in the data model today.

But again, in the implementation, we should only pick and support (in the code) a single one, today.

Reply:
Updated the design in 62be69907e2e8e3de9ca7e66a2987e39b13f6947

Reply:
For now given the timeline, I am going ahead with `AES256GCM` and will support `secretBox` in future implementations